design/cdk-bootstrap.md

```diff
@@ -0,0 +1,199 @@
+
+All of the above resources will be created with well-defined physical names -
+none of them will rely on automated CloudFormation naming.
+This is so that the other stages of CDK synthesis,
+like asset resolution, can rely on the concrete names
+(there is no reliable way to transfer this kind of information in an automated way across regions and/or accounts).
+This also allows for more fine-grained permissions -
+for instance, the continuous delivery pipeline needs to grant the **action role**
+permissions to read from the pipeline's bucket
+(to read the artifact that contains the CFN template to deploy);
+this way, it can add a well-defined name to the resource policy of the artifact bucket,
+instead of granting those permissions to all principals in the account.
+
+The naming scheme will include the following elements in order to minimize the chance of name collisions:
+
+* The region we're bootstrapping in.
+* The account ID we're bootstrapping in.
+* The type of the resource (file assets bucket, Docker assets repository, etc.).
+* An optional *qualifier* that defends against S3 bucket sniping.
+
+### CLI options in detail
+
+These options are inherited from the current CLI experience:
+
+* `--profile`: use the given local AWS credentials profile when interacting with the target environment.
+
+* `--stack-name`, `-s`: allows you to explicitly name the CloudFormation bootstrap stack
+  (instead of relying on the default naming scheme).
+
+* `--bucket-name`, `-b`: allows you to explicitly name the file assets S3 bucket
+  (instead of relying on the default naming scheme).
+
+* `--kms-key-id`, `-k`: optional identifier of the KMS key used for encrypting the file assets S3 bucket.
+
+These are the new options:
+
+* `--trust`: allows specifying an AWS account ID, or a list of them,
+  that the created roles (see above) should be assumable from.
+  This will be required to be passed as the pipeline account,
+  for deployment from a Continuous Delivery CDK pipeline to work.
+
+* `--qualifier`, `-q`: an optional qualifier added to the physical names of the bootstrap resources
+  in order to protect against name collisions, especially for things like S3 buckets
+  (which are globally unique in a given AWS partition).
+
+## Updating bootstrap resources
+
+Because we already have a bootstrap solution in place,
+and it's possible we will need to add more bootstrap resources as time goes by,
+we should have a mechanism in place for migrating,
+and giving meaningful errors if the bootstrapping has not been done for an environment that needs it.
+
+I don't think invoking the full `cdk-bootstrap` tool on every deploy is a good idea, though;
+I worry that calculating a full diff of actual versus desired resource state might impact the performance of commands like
+`deploy` too negatively.
+
+My proposal is to have an export on the bootstrap stack,
+called `ResourcesChecksum`, that will contain the SHA-256 hash of the state of all resources.
```

Comment:
It's a good idea but might be tricky since we can't implement this using "raw cloudformation" (which is a requirement). Perhaps just export a version number?

Also, at what stage in the deployment does this export get read? For example, when an app gets deployed through CodePipeline, at what stage do we make sure that the bootstrap resources are at the right version.

Reply:
Changed to a simple increasing version number, let me know if this is better.